File: app/src/main/java/net/vnnz/apps/kotlin/tracker/DataLoader.java
package net.vnnz.apps.kotlin.tracker;

import android.content.Context;
import android.support.v4.content.AsyncTaskLoader;

import net.vnnz.apps.kotlin.tracker.pojo.ListItem;
import net.vnnz.apps.kotlin.tracker.utils.ResourceUtils;

import java.util.ArrayList;
import java.util.List;

public class DataLoader extends AsyncTaskLoader<List<ListItem>> {

    int resId;

    public DataLoader(Context context, int resId) {
        super(context);
        this.resId = resId;
        onContentChanged();
    }

    @Override
    public List<ListItem> loadInBackground() {
        String jsonString = ResourceUtils.Companion.readJSONfromRaw(resId, getContext());
        try {
            Thread.sleep(3000);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        return new ArrayList<ListItem>();
    }

    @Override
    protected void onStartLoading() {
        if (takeContentChanged())
            forceLoad();
    }

    @Override
    protected void onStopLoading() {
        cancelLoad();
    }

}

File: app/src/main/java/net/vnnz/apps/kotlin/tracker/DataLoader.kt
package net.vnnz.apps.kotlin.tracker

import android.content.Context
import android.support.v4.content.AsyncTaskLoader

import com.google.gson.Gson

import net.vnnz.apps.kotlin.tracker.pojo.JSONResult
import net.vnnz.apps.kotlin.tracker.pojo.ListItem

import java.io.InputStreamReader
import java.util.ArrayList

class DataLoader(context: Context, internal var resId: Int) : AsyncTaskLoader<List<ListItem>>(context) {

    init {
        onContentChanged()
    }

    override fun loadInBackground(): List<ListItem> {
        val inputStream = context.resources.openRawResource(resId)
        val reader = InputStreamReader(inputStream)
        val response = Gson().fromJson(reader, JSONResult::class.java)
        val items = response.JSONobject?.items
        return items?.map(::ListItem) ?: ArrayList() // ::ListItem -> ListItem(it)
    }

    override fun onStartLoading() {
        if (takeContentChanged())
            forceLoad()
    }

    override fun onStopLoading() {
        cancelLoad()
    }

}